api/provider/aws/common.go

```diff
@@ -13,32 +14,61 @@ import (
 	"github.com/aws/aws-sdk-go/service/ecs/ecsiface"
 	"github.com/aws/aws-sdk-go/service/elb"
 	"github.com/aws/aws-sdk-go/service/elb/elbiface"
+	alb "github.com/aws/aws-sdk-go/service/elbv2"
+	albiface "github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
 	"github.com/quintilesims/layer0/api/tag"
 	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/retry"
 )
 
-func describeLoadBalancer(elbapi elbiface.ELBAPI, loadBalancerName string) (*elb.LoadBalancerDescription, error) {
-	input := &elb.DescribeLoadBalancersInput{}
-	input.SetLoadBalancerNames([]*string{aws.String(loadBalancerName)})
-	input.SetPageSize(1)
+// searches for the load balancer name as both classic and application load balancers and returns
+// the first found result or an error if the neither classic or application lb could be found for
+// the given lb name
+func describeLoadBalancer(elbapi elbiface.ELBAPI, albapi albiface.ELBV2API, loadBalancerName string) (*genericLoadBalancer, error) {
+	// search classic load balancers
+	elbInput := &elb.DescribeLoadBalancersInput{}
+	elbInput.SetLoadBalancerNames([]*string{aws.String(loadBalancerName)})
+	elbInput.SetPageSize(1)
 
-	output, err := elbapi.DescribeLoadBalancers(input)
+	if err := elbInput.Validate(); err != nil {
+		return nil, err
+	}
+
+	elbExists := true
+	elbOutput, err := elbapi.DescribeLoadBalancers(elbInput)
 	if err != nil {
-		if err, ok := err.(awserr.Error); ok && err.Code() == "LoadBalancerNotFound" {
-			return nil, errors.Newf(errors.LoadBalancerDoesNotExist, "LoadBalancer '%s' does not exist", loadBalancerName)
+		if err, ok := err.(awserr.Error); !ok || err.Code() != alb.ErrCodeLoadBalancerNotFoundException {
+			return nil, err
 		}
 
+		elbExists = false
+	}
+
+	if elbExists {
+		return newGenericLoadBalancer(elbOutput.LoadBalancerDescriptions[0], nil), nil
+	}
+
+	// search application load balancers
+	albInput := &alb.DescribeLoadBalancersInput{}
+	albInput.SetNames([]*string{aws.String(loadBalancerName)})
+
+	if err := albInput.Validate(); err != nil {
 		return nil, err
 	}
 
-	if len(output.LoadBalancerDescriptions) != 1 {
-		return nil, errors.Newf(errors.LoadBalancerDoesNotExist, "LoadBalancer '%s' does not exist", loadBalancerName)
+	albOutput, err := albapi.DescribeLoadBalancers(albInput)
+	if err != nil {
+		if err, ok := err.(awserr.Error); !ok || err.Code() == alb.ErrCodeLoadBalancerNotFoundException {
+			return nil, errors.Newf(errors.LoadBalancerDoesNotExist, "LoadBalancer '%s' does not exist", loadBalancerName)
+		}
+
+		return nil, err
```

Comment:
I understand the flow of the code here, but it seems awkward to execute it in this way. Would it be wrong to make calls to `elbapi.DescribeLoadBalancers(elbInput)` immediately followed by `albapi.DescribeLoadBalancers(albInput)` and then evaluate each of their respective `err`s after they return? 

It seems like you could cut down on the length of this logic, and you only need to check for LoadBalanacerDoesNotExist within one if block.

You would have something like 
```
if albErr != nil || elbErr != nil {
    <evaluate albErr.(awserr.Error) and elbErr.(awserr.Error)>

    return nil, errors.Newf(errors.LoadBalancerDoesNotExist, "LoadBalancer '%s' does not exist", loadBalancerName)
}
```

Reply:
I see where you're coming from, @jparsons04, and for my part, I think that here I have a slight preference for finding exit points sooner than later in favor of trying to make fewer API calls. It's a minor thing probably, but your proposal would have us _always_ making two `DescribeLoadBalancers()` calls; the current flow at least gives us the opportunity to sometimes only call it once if the load balancer happens to be a classic one.

I think this was @sesh-kebab's work - maybe he has a preference/reason he could weigh in with?

Reply:
> Would it be wrong to make calls to elbapi.DescribeLoadBalancers(elbInput) immediately followed by albapi.DescribeLoadBalancers(albInput) and then evaluate each of their respective errs after they return?

Short answer would be yes (or what tanner said). We don't want to make an additional api request if we don't have to.

Reply:
✅

Reply:
:heavy_check_mark: